Applications/Spire/Include/Spire/Spire/QtFuture.hpp

```diff
@@ -5,34 +5,66 @@
 #include <Beam/Routines/Async.hpp>
 #include <boost/noncopyable.hpp>
 #include "Spire/Spire/QtPromise.hpp"
-#include "Spire/Spire/Spire.hpp"
 
 namespace Spire {
 
   /*! \brief Encapsulates a QtPromise resolving to a future value.
       \tparam T The type of value to resolve to.
   */
   template<typename T>
-  class QtFuture : private boost::noncopyable {
+  class BaseQtFuture : private boost::noncopyable {
     public:
 
       //! The type of value to resolve to.
       using Type = T;
 
-      QtFuture(QtFuture&&) = default;
+      //! Resolves the QtPromise waiting on this future to an exception.
+      void resolve(std::exception_ptr e);
+
+    protected:
+      Beam::Routines::Eval<Type> m_eval;
+
+      BaseQtFuture(Beam::Routines::Eval<Type> eval);
+  };
+
+  /*! \brief Encapsulates a QtPromise resolving to a future value.
+      \tparam T The type of value to resolve to.
+  */
+  template<typename T>
+  class QtFuture : public BaseQtFuture<T> {
+    public:
+
+      QtFuture(QtFuture&&);
 
-      QtFuture& operator =(QtFuture&&) = default;
+      QtFuture& operator =(QtFuture&&);
+
+      using BaseQtFuture<T>::resolve;
 
       //! Resolves the QtPromise waiting on this future.
       void resolve(Type value);
 
-      //! Resolves the QtPromise waiting on this future to an exception.
-      void resolve(std::exception_ptr e);
+    private:
+      friend struct std::pair<QtFuture, QtPromise<Type>>;
```

Comment:
I didn't see the reason to friend `QtFuture` with the entire `std::pair` template, so I replaced it with the exact instantiation.
In fact, I would rather friend it with `make_future` function only and avoid piecewise construction altogether - the benefits are minuscule if any, and there can be some pitfalls regarding exception safety and such

Reply:
What exception safety issues are there? How would you implement ```make_future``` without inplace construction?

Reply:
I am not sure if there are any but the order of construction of the promise and the future is less defined, those being dependent on one another.

If I remember correctly, both `std::pair` constructor and `std::make_pair` work with rvalues for `first` and `second`. Even if there is a problem with `std::pair`, something like STL's syntax is also viable:
```
auto f = QtFuture<int>();
auto p = f.get_promise();
```
(STL's uses the opposite terminology for futures and promises.)

Reply:
STL's ```get_future``` is a significantly worse API that only exists because structured bindings weren't available back in C++11. As such the API isn't exception safe and also requires storing additional state.

If you can identify issues with the ```make_future``` implementation or API I am happy to hear them. As far as I know there is no other way to safely implement this.

Reply:
What's wrong with something like:
```
  template<typename T>
  std::pair<QtFuture<T>, QtPromise<T>> make_future() {
    auto async = std::make_unique<Beam::Routines::Async<T>>();
    auto eval = async->GetEval();
    auto f = QtFuture<T>(std::move(eval));
    auto p = QtPromise<T>([async = std::move(async)] {
      if constexpr(std::is_same_v<T, void>) {
        async->Get();
      } else {
        return async->Get();
      }
    }, LaunchPolicy::ASYNC))
    return std::make_pair(std::move(f), std::move(p));
  }
```

Reply:
For this use case nothing is particularly wrong with that since moves only involve pointers, if you want to modify the implementation to that go for it.

Reply:
To me, it is mostly a coding style concern. There are subtle differences between moving stuff in and in-place construction, which, under some circumstances, may affect exception safety and such, but most of the time they do not. Such details are quite difficult to navigate, let alone communicate. For the most part, the difference is all about compile error messages, which are far easier to navigate if no forwarding is made. In other words, it is the exact same concern as choosing vector push_back instead of emplace_back when the latter provides no performance benefits.

That said, if you preferred style is to construct pairs in-place, I will keep it as is. The only thing I am against is friending the class with the entire std::pair template, which is, ideologically, incorrect API to me: anyone specializing std::pair for any types (even something like `std::pair<int, int>`) can access private members of QtFuture. If someone specializes `std::pair<QtFuture<T>, QtPromise<T>>` and uses those members though, it seem fine as that programmer is aware of QtFuture at very least.